org/mozilla/jss/ssl/javax/JSSParameters.java

```diff
@@ -190,4 +198,27 @@ public String getHostname() {
     public void setHostname(String server_hostname) {
         hostname = server_hostname;
     }
+
+    public void setApplicationProtocols(String[] protocols) throws IllegalArgumentException {
```

Comment:
If you want to shake your fist at Java some more... :-)

[OpenJDK 9 b96](https://bugs.openjdk.java.net/browse/JDK-8144093) introduced ALPN support. We can only use JDK9+ on Fedora 33+. So we need something to support JDK8.

[OpenJDK 8 b252](https://bugs.openjdk.java.net/browse/JDK-8230977) should include ALPN support. However, this was only released late last year. We either introduce a dependency on a really recent version of the JDK (which we've historically avoided), or we introduce this into `JSSParameters`. Since `JSSParameters` exists to provide a mechanism for JSS-specific parameters to work (such as Hostname... </not-salty>) -- introducing another parameter for compatibility seems like a good idea.

My 2c but I'd rather not require a specific version of the JDK other than "at least JDK 8".

Reply:
@cipherboy thanks for clarifying.

FWIW f32 has OpenJDK 8 b265 and I confirmed it has the ALPN bits.

I don't mind if we reproduce the behaviour but there should at least be a comment that when we are depending on OpenJDK >=9 or >= 8 b252 then we can remove these implementations.

Reply:
We need to keep these anyways, as we need to convert to byte encoding. There are no internal helpers (on SSLEngine) the would let us access the advertised list anyways.

What could theoretically go away is the SSLParameters stuff, but again, that's useful for other reasons (converting to byte-encoding).

So that just leaves the JDKCompat shim as the only thing that can and should actually be removed in the future. IMO.